### ISSUE TITLE
Compare init method selector instead of name and number of arguments

### ISSUE THREAD

**Issue by jcarpanelli** [zeppelinos/zos#791]
ATM, when the user has an implementation contract that has more than one initializer function with the same number of arguments and tries to create a proxy selecting one of those methods,  `zos` fails with the following error:
```
Found more than one match for function initialize with 2 arguments in contract Foo
```
Contract example:

```solidity
contract Foo is Initializable {
  uint256 public namba;
  string public stringo;

  function initialize(uint256 _namba, string memory _stringo) public initializer {
    namba = _namba;
    stringo = _stringo;
  }

  function initialize(string memory _stringo, string memory _bar) public initializer {
    stringo = _stringo;
  }
}
```

To fix this, instead of comparing the method arguments length, compare the method selector or the method inputs by type.

**Comment by jachai** [comment]
Is there any workaround for this issue?

**Comment by jcarpanelli** [comment]
Hi @jachai, sorry for the late response! Unfortunately, The `cli` does not provide a workaround for this yet. There are some valid approaches to call a specific function though. For example, after creating the your contract instance (and thus having the contract instance address), you can do a super simple script using the programmatic library:

```js
'use strict';

const { Contracts, ZWeb3, Transactions, encodeCall } = require('zos-lib');
ZWeb3.initialize('http://localhost:7545');


async function callInitializerFunction() {
  const contractAddress = '0x0290FB167208Af455bB137780163b7B7a9a10C16'; // The already deployed instance
  const senderAddress = '0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1'; 
  const foo = 'foo';
  const buz = 'buz';
  const txParams = { from: senderAddress };
   
  // Send the transaction
  const data = encodeCall('initialize', ['string', 'string'], [foo, buz]);
  await Transactions.sendRawTransaction(contractAddress, { data }, txParams);
}

callInitializerFunction();

```

The `Transactions#sendRawTransaction` function will estimate both `gas` and `gasPrice` for you. 

As another alternative, you can use the `web3#Contract` API. Just replace the last 2 lines of the snippet above with the following:

```js
  const Bar = Contracts.getFromLocal('Bar').at(contractAddress);
  await Bar.at(contractAddress).methods['0x4cd88b76'](foo, buz).send(txParams);
``` 
Where `0x4cd88b76` are the first 4 bytes of the keccak-256 hash of the function.

**Comment by spalladino** [comment]
@jcarpanelli going back to this issue, I'd expect that if the user chooses the method from an interactive list of functions, that the `value` of that choice is the fully qualified method name (as in `initialize(uint256,string)`), so then the `ABI#buildCallData` knows which method to choose: 
https://github.com/zeppelinos/zos/blob/4c2900ac3af6fd0a911c4bfeadd40631846102d7/packages/lib/src/utils/ABIs.ts#L53-L59

Also, we could allow the user to specify `zos create MyContract --init initialize(uint256,string)` in the command line, to choose exactly which method to call. WDYT?

**Comment by jcarpanelli** [comment]
Hey @jachai! This fix was released a couple of weeks ago in `zos@2.4`. I recommend you to try it out! I'm closing this issue, but feel free to reopen it if necessary!
